### PR TITLE
gradient verification callback

### LINKED ISSUE #194: Some callbacks for debugging and visualization
## 🚀 Feature

Do you have interest in adding any of these callbacks here? 
https://github.com/awaelchli/pytorch-lightning-snippets
Can also make an issue for each one, just wanted to know first if there is a desire for more callbacks in PL bolts?

[TrainingDataMonitor](https://github.com/awaelchli/pytorch-lightning-snippets#monitoring-training-data): logs histogram of what goes into training_step
[ModuleDataMonitor](https://github.com/awaelchli/pytorch-lightning-snippets#monitoring-in--and-outputs-of-network-layers): logs histogram of what goes into each submodule in LightningModule
[BatchNormVerification](https://github.com/awaelchli/pytorch-lightning-snippets#normalization-layers-and-biases): checks (heuristic) if you have a bias before a normalization layer
[BatchGradientVerification](https://github.com/awaelchli/pytorch-lightning-snippets#mixing-data-across-the-batch-dimension): checks that your network does not mix data across batch dim.

### Motivation

- Have more callbacks in PL bolts
- useful for debugging and visualization of training

### Additional context

Happy to help maintain it and adapt to your needs.

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Fixes #194 
Adds the requested callbacks for model verification. 
It's copied from [my personal repository](https://github.com/awaelchli/pytorch-lightning-snippets#mixing-data-across-the-batch-dimension).

cc @TylerYep 

## Before submitting
- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning-bolts/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests?
- [x] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning-bolts/blob/master/CHANGELOG.md)?

## PR review
 - [x] Is this pull request ready for review? (if not, please submit in draft mode)

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Review by Borda — APPROVED** [review]
LGTM, just would be nice to have tests for uncovered lines...

**Comment by awaelchli** [comment]
@Borda ok, I will try to make it higher

**Comment by TylerYep** [comment]
Thank you so much for implementing this!

File: docs/source/info_callbacks.rst [hunk 1]
```diff
@@ -64,3 +64,63 @@ You can track all or just a selection of submodules:
...
+|
+
+If the gradient is non-zero for the other samples in the batch, it means the forward pass of the model is mixing data!
+The :class:`~pl_bolts.callbacks.verification.batch_gradient.BatchGradientVerificationCallback`
+does all of that for you before training begins.
+
+.. code-block:: python
+
+    from pytorch_lightning import Trainer
+    from verification.batch_gradient import BatchGradientVerificationCallback
```
Comment (akihironitta) [inline docs/source/info_callbacks.rst:98]:
```suggestion
    from pl_bolts.callbacks import BatchGradientVerificationCallback
```

File: docs/source/info_callbacks.rst [hunk 2]
```diff
@@ -64,3 +64,63 @@ You can track all or just a selection of submodules:
...
+    Check the operations that reshape and permute tensor dimensions in your model.
+
+
+A non-Callback version
+:class:`~pl_bolts.callbacks.verification.batch_gradient.BatchGradientVerification`
+that works with any PyTorch :class:`~torch.nn.Module` is also available:
+
+.. code-block:: python
+
+    from verification.batch_gradient import BatchGradientVerification
```
Comment (akihironitta) [inline docs/source/info_callbacks.rst:120]: Shall we make `BatchGradientVerification` available right under `pl_bolts.callbacks`? (although the class is not a `Callback` of PL.) or any other place to have the class in Bolts in your mind?
```suggestion
    from pl_bolts.callbacks import BatchGradientVerification
```

File: pl_bolts/callbacks/verification/base.py [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+
+class VerificationCallbackBase(Callback):
+    """
+    Base class for model verification in form of a callback.
+    This type of verification is expected to only work with
+    :class:`~pytorch_lightning.core.lightning.LightningModule` and will take the input array
+    from :attr:`~pytorch_lightning.core.lightning.LightningModule.example_input_array` if needed.
+    """
+
+    def __init__(self, warn: bool = True, error: bool = False):
```
Comment (akihironitta) [inline pl_bolts/callbacks/verification/base.py:91]: Let's add `-> None` as [PEP 484](https://www.python.org/dev/peps/pep-0484/#the-meaning-of-annotations) suggests:
> Note that the return type of `__init__` ought to be annotated with `-> None`.

```suggestion
    def __init__(self, warn: bool = True, error: bool = False) -> None:
```

File: pl_bolts/callbacks/verification/base.py [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+        """
+        Arguments:
+            warn: If `True`, prints a warning message when verification fails. Default: `True`.
+            error: If `True`, prints a error message when verification fails. Default: `False`.
```
Comment (akihironitta) [inline pl_bolts/callbacks/verification/base.py:95]:
```suggestion
            warn: If ``True``, prints a warning message when verification fails. Default: ``True``.
            error: If ``True``, prints an error message when verification fails. Default: ``False``.
```

File: pl_bolts/callbacks/verification/base.py [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+        """
+        self._raise_warning = warn
+        self._raise_error = error
+
+    def message(self, *args, **kwargs) -> str:
```
Comment (akihironitta) [inline pl_bolts/callbacks/verification/base.py:100]:
```suggestion
    def message(self, *args: Any, **kwargs: Any) -> str:
```

File: pl_bolts/callbacks/verification/base.py [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+        Arguments:
+            *args: Any positional arguments that are needed to construct the message.
+            **kwargs: Any keyword arguments that are needed to construct the message.
+
+        Returns:
+            The message as a string.
+        """
+        pass
+
+    def warning_message(self, *args, **kwargs) -> str:
```
Comment (akihironitta) [inline pl_bolts/callbacks/verification/base.py:116]:
```suggestion
    def warning_message(self, *args: Any, **kwargs: Any) -> str:
```

File: pl_bolts/callbacks/verification/base.py [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+        """ The warning message printed when the model does not pass the verification. """
+        return self.message(*args, **kwargs)
+
+    def error_message(self, *args, **kwargs) -> str:
```
Comment (akihironitta) [inline pl_bolts/callbacks/verification/base.py:120]:
```suggestion
    def error_message(self, *args: Any, **kwargs: Any) -> str:
```

File: pl_bolts/callbacks/verification/base.py [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+        """ The error message printed when the model does not pass the verification. """
+        return self.message(*args, **kwargs)
+
+    def _raise(self, *args, **kwargs):
```
Comment (akihironitta) [inline pl_bolts/callbacks/verification/base.py:124]:
```suggestion
    def _raise(self, *args: Any, **kwargs: Any) -> None:
```

File: pl_bolts/callbacks/verification/batch_gradient.py [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+    This can happen if reshape- and/or permutation operations are carried out in the wrong order or
+    on the wrong tensor dimensions.
+    """
+
+    def check(
+        self,
+        input_array: Any,
+        input_mapping: Optional[Callable] = None,
+        output_mapping: Optional[Callable] = None,
+        sample_idx=0,
```
Comment (akihironitta) [inline pl_bolts/callbacks/verification/batch_gradient.py:22]:
```suggestion
        sample_idx: int = 0,
```

File: pl_bolts/callbacks/verification/batch_gradient.py [hunk 10]
```diff
@@ -0,0 +1,199 @@
...
+                (tuples, lists, dicts) of tensors and combine them automatically. See
+                :func:`default_output_mapping` for more information on the default behavior.
+            sample_idx:
+                The index `i` of the batch sample to run the test for. When computing the gradient of
+                a loss value on the `i-th` output w.r.t. the whole input, we expect the gradient to be
+                non-zero only on the `i-th` input sample and zero gradient on the rest of the batch.
+                Default: `i = 0`.
+
+        Returns:
+             `True` if the data in the batch does not mix during the forward pass, and `False` otherwise.
```
Comment (akihironitta) [inline pl_bolts/callbacks/verification/batch_gradient.py:46]:
```suggestion
             ``True`` if the data in the batch does not mix during the forward pass, and ``False`` otherwise.
```

File: pl_bolts/callbacks/verification/batch_gradient.py [hunk 11]
```diff
@@ -0,0 +1,199 @@
...
+    Verification is performed right before training begins.
+    """
+
+    def __init__(
+        self,
+        input_mapping: Callable = None,
+        output_mapping: Callable = None,
+        sample_idx=0,
+        **kwargs
+    ):
```
Comment (akihironitta) [inline pl_bolts/callbacks/verification/batch_gradient.py:93]:
```suggestion
        input_mapping: Optional[Callable] = None,
        output_mapping: Optional[Callable] = None,
        sample_idx: int = 0,
        **kwargs: Any,
    ):
```

File: pl_bolts/callbacks/verification/batch_gradient.py [hunk 12]
```diff
@@ -0,0 +1,199 @@
...
+            sample_idx: The index of the batch sample to run the test for.
+                See :meth:`BatchGradientVerification.check` for more information.
+            **kwargs: Additional arguments for the base class :class:`VerificationCallbackBase`
+        """
+        super().__init__(**kwargs)
+        self._input_mapping = input_mapping
+        self._output_mapping = output_mapping
+        self._sample_idx = sample_idx
+
+    def message(self):
```
Comment (akihironitta) [inline pl_bolts/callbacks/verification/batch_gradient.py:109]:
```suggestion
    def message(self) -> str:
```

File: pl_bolts/callbacks/verification/batch_gradient.py [hunk 13]
```diff
@@ -0,0 +1,199 @@
...
+        message = (
+            "Your model is mixing data across the batch dimension."
+            " This can lead to wrong gradient updates in the optimizer."
+            " Check the operations that reshape and permute tensor dimensions in your model."
+        )
+        return message
+
+    def on_train_start(self, trainer, pl_module):
```
Comment (akihironitta) [inline pl_bolts/callbacks/verification/batch_gradient.py:117]:
```suggestion
    def on_train_start(self, trainer: Trainer, pl_module: LightningModule) -> None:
```

File: pl_bolts/callbacks/verification/batch_gradient.py [hunk 14]
```diff
@@ -0,0 +1,199 @@
+from typing import Any, Callable, List, Optional
+
+import torch
```
Comment (akihironitta) [inline pl_bolts/callbacks/verification/batch_gradient.py:3]: For type hints of `BatchGradientVerificationCallback`:
```suggestion
import torch
from pytorch_lightning import LightningModule, Trainer
```

**Review by akihironitta — CHANGES_REQUESTED** [review]
@awaelchli I love this feature! I left some minor suggestions on typing and docs, so would you mind having a look at them? They should be resolved by clicking some buttons in your browser :]

File: docs/source/info_callbacks.rst (hunk 2 above)
Comment (awaelchli) [reply docs/source/info_callbacks.rst:120]: I was wondering the same. We don't want it to show up automatically in the docs under callbacks though. Maybe the utils folder is another options?
We could add from `pl_bolts.callbacks.verification import BatchGradientVerification` to the `__init__` in utils and still keep the code together (from pl_bolts.utils import BatchGradientVerification).

**Comment by tchaton** [comment]
Hey @awaelchli, why not push it directly to PL ?

**Comment by awaelchli** [comment]
Because @williamFalcon wants only essential callbacks in core PL library, and all non-essential, specialized callbacks here in bolts. Or at least this the information I have back from when pl_bolts was created, maybe it changed. I do not want to force this into any library unless team/community wants it.

**Comment by tchaton** [comment]
> Because @williamFalcon wants only essential callbacks in core PL library, and all non-essential, specialized callbacks here in bolts. Or at least this the information I have back from when pl_bolts was created, maybe it changed. I do not want to force this into any library unless team/community wants it.

This looks like a great feature and we need to provide more callback examples to people.

It is up to you, but I see its value for PL.

**Comment by awaelchli** [comment]
> This looks like a great feature and we need to provide more callback examples to people.
> It is up to you, but I see its value for PL.

Thank you for the kind feedback. Ok, let me adress the remaining review comments here on the PR and I will double check with William.

File: pl_bolts/callbacks/verification/base.py [hunk 15]
```diff
@@ -0,0 +1,126 @@
...
+    def __init__(self, model: nn.Module):
+        """
+        Arguments:
+            model: The model to run verification for.
+        """
+        super().__init__()
+        self.model = model
+
+    @abstractmethod
+    def check(self, *args: Any, **kwargs: Any) -> bool:
```
Comment (awaelchli) [inline pl_bolts/callbacks/verification/base.py:26]: @akihironitta this mypy tool is not very smart :( It is not liking that the subclass has different args than here. I want this abstract method to be as general as possible and not specify concrete arguments and types. It should only act as an interface. Any suggestions how to proceed? I believe I have to add `# type: ignore`

File: docs/source/info_callbacks.rst (hunk 2 above)
Comment (akihironitta) [reply docs/source/info_callbacks.rst:120]: That sounds reasonable, I suppose.

File: pl_bolts/callbacks/verification/base.py (hunk 15 above)
Comment (akihironitta) [reply pl_bolts/callbacks/verification/base.py:26]: I found a related issue in mypy repo, and it basically says that an easy workaround would be to add `# type: ignore[override]`.
https://github.com/python/mypy/issues/1237#issuecomment-579895019

Comment (awaelchli) [reply pl_bolts/callbacks/verification/base.py:26]: no chance, it won't work. I tried to put it everywhere: at the top of method, on the same line as signature, below it, on top of the class, on top of the file, both in subclass and superclass.

Comment (awaelchli) [reply pl_bolts/callbacks/verification/base.py:27]: I had to spam `#type: ignore` everywhere to make it work. 
This mypy tool, I don't understand it. I spent hours now studying the docs of this tool trying to figure out what the error messages mean. I tried everything, but the `type: ignore` are unavoidable, yet they pollute the code unnecessarily. It's unbelievably frustrating, and I can no longer work on this, sorry.

Comment (akihironitta) [reply pl_bolts/callbacks/verification/base.py:27]: I completely understand your frustration. Let's ignore them all for now.

**Review by akihironitta — APPROVED** [review]

File: pl_bolts/callbacks/verification/batch_gradient.py [hunk 16]
```diff
@@ -0,0 +1,193 @@
...
+            output_mapping: An optional output mapping that combines all batched tensors in the output
+                collection into one big batch of shape (B, N), where N is the total number of dimensions
+                that follow the batch dimension in each tensor. By default, we handle nested collections
+                (tuples, lists, dicts) of tensors and combine them automatically. See
+                :func:`default_output_mapping` for more information on the default behavior.
+            sample_idx:
+                The index `i` of the batch sample to run the test for. When computing the gradient of
+                a loss value on the `i-th` output w.r.t. the whole input, we expect the gradient to be
+                non-zero only on the `i-th` input sample and zero gradient on the rest of the batch.
+                Default: `i = 0`.
```
Comment (Borda) [inline pl_bolts/callbacks/verification/batch_gradient.py:45]:
```suggestion
```

File: tests/callbacks/verification/test_batch_gradient.py [hunk 17]
```diff
@@ -0,0 +1,260 @@
...
+    assert result == is_valid
+
+
+@pytest.mark.parametrize(
+    "gpus",
+    [
+        pytest.param(0),
+        pytest.param(
+            1,
+            marks=pytest.mark.skipif(not torch.cuda.is_available(), reason="Test requires GPU"),
```
Comment (Borda) [inline tests/callbacks/verification/test_batch_gradient.py:139]:
```suggestion
            marks=pytest.mark.skipif(not torch.cuda.is_available(), reason="Test requires GPU")
```
